docs/type-annotations.org

```diff
@@ -0,0 +1,219 @@
+* Type annotations
+
+In Elisp users are not required to provide type annotations to their
+code.  While at many places the types can be inferred there are
+places, especially in user-defined functions, where we can not infer
+the correct type.
+
+Therefore Elsa provides users with the option to annotate their
+function definitions.  The annotations are placed in the body of the
+function inside a =declare= form:
+
+#+BEGIN_SRC elisp
+(defun add-one (x)
+  (declare (elsa (int) int))
+  (1+ x))
+#+END_SRC
+
+The =declare= form starts with =elsa= followed by one or two values.  If
+two are provided, first must be a list which is the list of input
```

Comment:
We might want to drop the parens around arguments and just have a sequence of types where the last one is automatically return type.

Reply:
Hm, that would be awkward in case you want a void function or rather, ignore the return value. Requiring parentheses around arguments would model this naturally (nothing would follow them) whereas without them you'd always need to specify the last thing. Also, what if you have an empty argument list, but a meaningful return value (like, `emacs-version`)? How would you tell both apart without parentheses around the arguments? If you need a better separator of arguments and return value, I like `:` in the Wirth family, an arrow would be another option.

Reply:
You can always inspect the argument list, so incase of `(point)` you know the `int` is the return value since there is no argument.

I have never considered void functions, such a thing does not really exist in Elisp does it?  At the very least you always return `nil` or some such.

Reply:
Of course there are void functions, everything you call for its side effect, like `save-buffer` (returns `nil`) or `revert-buffer` (returns `t`). What I'm pointing out is that we could avoid ambiguity with sexp-syntax.

Reply:
A void function returns nothing.  Every function in elisp returns something.

Doing `(string-to-number (save-buffer))` would error.  We don't like errors, therefore we must always consider the return type... the above code is silly but might be a result of a failed refactoring or a mistake and we want to emit a warning there.

Reply:
how about smth like
```
(defun add-one (x)
  (declare (elsa-> int int))
  (1+ x))

(defun add-one-and-debug! (x msg)
  (declare (elsa-> str int int))
  (message "%s" msg)
  (1+ x))
```

Reply:
how about smth like
```
(defun add-one (x)
  (declare (elsa-> int int))
  (1+ x))

(defun add-one-and-debug! (x msg)
  (declare (elsa-> str int int))
  (message "%s" msg)
  (1+ x))
```

Reply:
Void is nasty because it doesn't play well with the type algebra.

What would this function return?

```elisp
(defun what-is-return-type ()
  (if (we-should-save-buffer) ;; bool
      (save-buffer) ;; void
    (return-time-since-last-save-in-seconds) ;; int
    ))
```

A type like `void | int` makes no sense because void basically poisons everything.  If you would want to use such a function you would have to test the return type anyway (did I get an int back?). So then the type of `save-buffer` doesn't really matter.

It's a bit silly example, but the problem is that void is not a type in a sense every other type is a type (i.e. a set).  Void is not even an empty set type, it's something outside the hierarchy.

This function 

```elisp
(defun forever-and-ever ()
  (while t t))
```

also returns void?  Or is that something else.

I have this issue https://github.com/emacs-elsa/Elsa/issues/68 to check for unused return values so I would rather go with a "this function should be only used for side-effect" warning but not on the type level.

Reply:
You are much more knowledgeable than me in this area. Do what you think is best!

Reply:
> A type like void | int makes no sense because void basically poisons everything. 

Some languages (Rust, OCaml) have unit types for this purpose `()`.  In this example, `(or void int)` tells you that you may get a useful value (`int`) or you may get a useless one (`void`).  You *have* to test the return type in order to use the `int`, because you cannot do anything with `void`.  But that's the point of the type analysis: to make sure your code is correct in all situations.  

> If you would want to use such a function you would have to test the return type anyway (did I get an int back?). So then the type of save-buffer doesn't really matter.

But it *does* matter.  Consider the alternative: `(or bool int)`, in which case you may freely use the `bool` return value even though it's undocumented, and may change in the future as argued by @DamienCassou.  If it changes to `int`, then if you previously made use of the `bool` value your code won't typecheck anymore, but there *is* value in an annotation to indicate "this value is not relevant and subject to change".

> This function
(defun forever-and-ever ()
  (while t t))
also returns void? Or is that something else.

Something else: it's the empty type (or bottom type).  It indicates that this function *never returns*, which is different than `void` as argued above, which means *returns something you cannot use*.

Reply:
To complete: the empty type may be more tricky to deal with, but in practice may not be that useful.  The `void` unit type however, though by not mean indispensable. is useful.